[functorch][nn] Refactor NN stateless APIs by swapping module tensors

XuehaiPan opened the pull request:
- Fixes #92295
- Resolves #86708
- Resolves #92153
- Closes #92401
- Closes #92218

- Requires #91579

Refactor NN stateless APIs by swapping module tensors.

cc @albanD @zou3519

XuehaiPan committed:
Cache intermediate submodules for submodule lookup in `functional_call`

XuehaiPan committed:
Cache intermediate submodules for submodule lookup in `make_functional`

XuehaiPan committed:
Extract reusable code to `NamedMemberAccessor`

XuehaiPan committed:
Refactor NN stateless APIs by swapping module tensors

jbschlosser commented:
Hey @XuehaiPan, thanks for the contribution! This is certainly something we'd consider merging. Do you mind answering a few questions to help our understanding on the review side?

* Can you describe at a somewhat detailed level how this approach works and differs from what is currently done in master?
* Are there benefits to this approach beyond addressing the issues linked in the PR description (which is of course very valuable)?
* What are the observable differences from the user side (beyond bug fixes)? Do you have a sense for where BC-breaking aspects will arise?
* Do you expect any performance improvements or regressions?

Apologies in advance if it takes us a while to review this; we're a bit busy on our side, but this is something we will work to get reviewed!

XuehaiPan commented:
> Can you describe at a somewhat detailed level how this approach works and differs from what is currently done in master?

I have described the implementation details of `functorch.make_functional` and (current approach) `torch.func.functional_call` in https://github.com/pytorch/pytorch/issues/92295#issue-1536084573.

I can summarize them as:

- `functorch.make_functional`: swap module tensors by `delattr` first when `setattr`. That will make all tensors store in `module._buffers` and leave an empty `module._paramters`.

    While doing a functional call, for example, given `name='weight'` and `tensor=Tensor(..., requires_grad=True)`, the `delattr` function will unregister the stateless module parameter, then `setattr` will register the user-given tensor as a buffer, not a parameter. Because `nn.Module.__setattr__` only set `nn.Parameter` type as a parameter.

- `torch.func.functional_call`: change the module class by overriding the `__getattribute__` and `__setattr__` methods. That does not change contents in  `module._parameters` and `module._buffers`.

    Current `torch.func.functional_call` implementation is very tricky. That allows `self.weight` access to the user-given tensor with a customized `__getattribute__` method. However, it does not change the contents in  `module._parameters` and `module._buffers`. So users can access the parameters/buffers by `self.<attr-name>` but all other methods are not available. Such as, `self.get_parameter(...)`, `self.parameters()`, `self.buffer()`, `self.state_dict()`, etc. These methods will return the tensors in the original module rather than the user-given ones.

Both of the existing two approaches above limit the usage and flexibility of user-defined forward pass. Users can only access the parameters/buffers by `self.<attr-name>`. If users use something else but available in NN APIs, users will get errors or unexpected values.

In this PR, the new approach is more like the `functorch.make_functional` approach (actually the `_swap_state` function), but not using `delattr + setattr`. Instead, it swaps the tensors in the module by accessing the `module._parameters` and `module._buffers` dicts directly. All user-given tensors are registered as _"true"_ parameters and buffers. Users can access them by `self.<attr-name>`, `self.get_parameter(...)`, `self.parameters()`, `self.state_dict()`, and also `self.to(...)`.

It is similar to the following for easy understanding (only similar in concept but not the same):

```python
def functional_call(module, parameters_and_buffers, args, kwargs):
    orig_state_dict = module.state_dict()
    try:
        module.load_state_dict(parameters_and_buffers)
        return module(*args, **kwargs)
    finally:
        module.load_state_dict(orig_state_dict)
```

------

> Are there benefits to this approach beyond addressing the issues linked in the PR description (which is of course very valuable)?

All user-given tensors are registered as _"true"_ parameters and buffers. Users can access them by `self.<attr-name>`, `self.parameters()`, `self.state_dict()`, and also `self.to(...)`. This will give the user a good experience transitioning from the object-originated pattern (OOP) to the stateless functional pattern. All methods, that the NN users are familiar with, are still available during the functional call.

> What are the observable differences from the user side (beyond bug fixes)? Do you have a sense for where BC-breaking aspects will arise?

No. It is feeless from the user side. Beyond bug fixes, it is backward-compatible.

> Do you expect any performance improvements or regressions?

I think it will have a similar performance to the current `torch.func.functional_call` implementation.

In my opinion, the performance would be:

call-per-second (the higher the better): `functorch.make_functional` > `torch.func.functional_call` (this PR) ≈ `torch.func.functional_call` (in master)

Because `functorch.make_functional` traverse the module and extract the tensor names in the first place. While `torch.func.functional_call` will traverse and do many checks in each function call. I think the former one would be slightly faster. But the true bottleneck here is the forward pass rather than the call preparation overhead though.

zou3519 commented:
Just a quick fyi @XuehaiPan -- I'll be out until 3/7 so probably won't respond very much. You can cc Chillee for general functorch matters and he'll route them to the right places

ain-soph commented:
Shall we add some unittest for the new `strict` flag?

XuehaiPan committed:
Update key names handling

XuehaiPan committed:
Add deprecation warnings

XuehaiPan committed:
Refactor weight tying

XuehaiPan committed:
Add functional call tests for strict flag

XuehaiPan committed:
Add functional call tests for member reference

XuehaiPan committed:
Add more tests

jbschlosser commented on torch/nn/utils/stateless.py:
```diff
@@ -37,144 +45,169 @@ def _setattr(self, name: str, value: Any) -> None:
+            )
+        if len(error_msgs) > 0:
+            raise RuntimeError(
+                "Error(s) in reparameterizing for {}:\n\t{}".format(
```
```suggestion
                "Error(s) in reparametrizing for {}:\n\t{}".format(
```

XuehaiPan replied:
I'm fine with both `reparameterize` and `reparametrize`. Both spellings are correct and both are used in PyTorch. E.g., `torch.distributions` uses "parameterize" (/w "e") while in `torch.nn` and `torch.aot_autograde`, we use "parametrize" (/wo "e").

"parameterize" (/w "e") looks more common.

<img width="1270" alt="image" src="https://user-images.githubusercontent.com/16078332/216742681-ab8976ab-8091-4e75-89fc-9353703c5dbb.png">

albanD replied:
The general "rule" we use for this (to try and preserve consistency):
- parameterize means turning something into a parameter.
- parametrize means reformulating something into something else via parameters (parametrize a curve).

Not that these are necessarily correct but at least that allows us to (try) to be consistent within PyTorch).

jbschlosser commented on torch/nn/utils/stateless.py:
```diff
@@ -37,144 +45,169 @@ def _setattr(self, name: str, value: Any) -> None:
-    module: 'torch.nn.Module',
+    module: "torch.nn.Module",
     parameters_and_buffers: Dict[str, Tensor],
+    *,
```
I think making `tie_weights` a kw-only arg is technically BC-breaking, since it affects the signature of this function which is imported as public into `torch.nn.utils._stateless` (see [here](https://github.com/pytorch/pytorch/blob/e52786f3d177a7ca5d490a516cf52e236ef072cb/torch/nn/utils/_stateless.py#L15)).

That said, I don't think we care too much since `_stateless` is deprecated. cc @albanD wdyt? when do we plan to stop caring about `_stateless` BC?

XuehaiPan replied:
I think this change is minor because `tie_weights` defaults to be `True`. If users do not want to tie weights, they can simply pass a full `parameters_and_buffers` dict rather than set `tie_weights=False`.

albanD replied:
It has been long enough IIRC and we should consider removing `_stateless` altogether.

XuehaiPan replied:
I removed `torch.nn.utils._stateless` in the newest commit. It has been deprecated for 10 months.

jbschlosser replied:
@XuehaiPan, do you mind bumping the removal of `torch.nn.utils._stateless` to a follow-up PR? I think we want to keep the scope of this PR to the refactor. Also, we generally like to make hard-breaking changes in isolation so they're easier to revert if needed.

XuehaiPan replied:
I reverted the breaking change of the file removal and kwonly `tie_weights`.

jbschlosser commented on test/test_stateless.py:
```diff
@@ -220,6 +221,197 @@ def test_reparamertize_module_fail_reset_to_original(self, functional_call):
+        subtest(torch.func.functional_call, "torch_func"),
+        subtest(stateless.functional_call, "stateless")
+    ])
+    def test_reparamertize_some_weights(self, functional_call):
```
```suggestion
    def test_reparametrize_some_weights(self, functional_call):
```

And throughout also - I realize this typo was present before this PR. Should be okay to rename them; none of the tests are flaky so nothing depends on their current name.

jbschlosser commented on test/test_stateless.py:
```diff
@@ -220,6 +221,197 @@ def test_reparamertize_module_fail_reset_to_original(self, functional_call):
         self.assertTrue('l1.parametrizations.weight.original' in dict(module.named_parameters()))
         self.assertEqual(orig_sn_weight, module.l1.weight)
 
+    @parametrize("functional_call", [
+        subtest(torch.func.functional_call, "torch_func"),
+        subtest(stateless.functional_call, "stateless")
+    ])
```
Nice, I like the coverage of hitting both `functional_call()` impls.

jbschlosser commented on test/test_stateless.py:
```diff
@@ -220,6 +221,197 @@ def test_reparamertize_module_fail_reset_to_original(self, functional_call):
+            re.escape("Missing key(s): 'buffer', 'l1.bias'."),
+        ):
+            out = functional_call(module, parameters, x, strict=True)
+        self.assertTrue(
+            all(t1 is t2 and torch.allclose(t1, t3)
+                for t1, t2, t3
+                in zip(orig_tensors,
+                       tuple(module.parameters()) + tuple(module.buffers()),
+                       orig_tensors_values)),
+            'the module should not have been modified by a failed call',
+        )
```
maybe pull this out to a helper function to reduce duplication throughout?

jbschlosser reviewed (commented):
Thanks for the hard work on this! Just made a pass and found it to be really solid.

Test coverage looks good; I see cases for:
* Calling `self.parameters()` in the forward pass (#92295)
* Weight tying support (#86708)
* Strict mode (#92153)

I made a few very minor comments, including a single possible BC-breaking thing (pretty niche; likely not important), but overall it looks great!

XuehaiPan committed:
Fix typos

XuehaiPan committed:
Extract reusable function

XuehaiPan committed:
Use `collections.Counter`

XuehaiPan committed:
Remove deprecated `torch.nn.utils._stateless`

XuehaiPan committed:
Revert "Remove deprecated `torch.nn.utils._stateless`"

This reverts commit 79449e678d17795a94f78c62971c41d09f52b19a.

XuehaiPan committed:
Revert kwonly for `tie_weights`

jbschlosser reviewed (approved):
Beautiful :)

jbschlosser commented:
@pytorchbot merge
